### PR TITLE
Feat/add multiple threshold #379

### PR DESCRIPTION
**Pull request by pbmiguel**
Me and @FranciscoSousaDeveloper tried to continue this Pull Request but we had no permission. So this Pull Request is the continuation of #385 with rebase and updated documentation.

Basically, this Pull Request enables the specification of multiple threshold values.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/coverlet-coverage/coverlet?pullRequest=1123) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/coverlet-coverage/coverlet?pullRequest=1123) before we can accept your contribution.<br/><br/>:x: pbmiguel [sign now](https://cla.dotnetfoundation.org/coverlet-coverage/coverlet?pullRequest=1123)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/coverlet-coverage/coverlet?pullRequest=1123) it.</sub>

**Comment by pbmiguel** [comment]
@MarcoRossignoli can you please validate?

File: src/coverlet.core/CoverageResult.cs [hunk 1]
```diff
@@ -149,22 +149,23 @@ public ThresholdTypeFlags GetThresholdTypesBelowThreshold(CoverageSummary summar
                         double line = summary.CalculateLineCoverage(Modules).AverageModulePercent;
                         double branch = summary.CalculateBranchCoverage(Modules).AverageModulePercent;
                         double method = summary.CalculateMethodCoverage(Modules).AverageModulePercent;
+                        var numModules = Modules.Count;
```
Comment (MarcoRossignoli) [inline src/coverlet.core/CoverageResult.cs:152]: We already got the avg we don't need to divide by num of modules. We're changing old behavior.

File: src/coverlet.console/Program.cs [hunk 2]
```diff
@@ -174,20 +176,48 @@ static int Main(string[] args)
...
+                {
+                    var thresholdValues = threshold.Value().Split(',', StringSplitOptions.RemoveEmptyEntries).Select(t => t.Trim());
+                    if (thresholdValues.Count() != thresholdTypeFlagQueue.Count())
+                    {
+                        throw new Exception($"Threshold type flag count ({thresholdTypeFlagQueue.Count()}) and values count ({thresholdValues.Count()}) doesnt match");
+                    }
+
+                    foreach (var thresholdValue in thresholdValues)
+                    {
+                        thresholdTypeFlagValues[thresholdTypeFlagQueue.Dequeue()] = double.Parse(thresholdValue);
```
Comment (MarcoRossignoli) [inline src/coverlet.console/Program.cs:211]: We should use `double.TryParse` and output meaniful message

File: src/coverlet.msbuild.tasks/CoverageResultTask.cs [hunk 3]
```diff
@@ -125,24 +125,52 @@ public override bool Execute()
...
+                {
+                    var thresholdValues = Threshold.Split(new char[] {','}, StringSplitOptions.RemoveEmptyEntries).Select(t => t.Trim());
+                    if(thresholdValues.Count() != thresholdTypeFlagQueue.Count())
+                    {
+                        throw new Exception($"Threshold type flag count ({thresholdTypeFlagQueue.Count()}) and values count ({thresholdValues.Count()}) doesnt match");
+                    }
 
+                    foreach (var threshold in thresholdValues)
+                    {
+                        thresholdTypeFlagValues[thresholdTypeFlagQueue.Dequeue()] = double.Parse(threshold);
```
Comment (MarcoRossignoli) [inline src/coverlet.msbuild.tasks/CoverageResultTask.cs:160]: ditto `double.TryParse`

**Review by MarcoRossignoli — CHANGES_REQUESTED** [review]
Some changes and here we go!
Thanks!

File: src/coverlet.core/CoverageResult.cs (hunk 1 above)
Comment (pbmiguel) [reply src/coverlet.core/CoverageResult.cs]: removed numModules. can you please check?

File: src/coverlet.console/Program.cs (hunk 2 above)
Comment (pbmiguel) [reply src/coverlet.console/Program.cs]: changed as suggested. please check

**Comment by MarcoRossignoli** [comment]
@pbmiguel seems that rebase didn't go well...I see 22 files changed you've included some old commits.

**Comment by pbmiguel** [comment]
@MarcoRossignoli, it's fixed now.

You can see the new changes [here](https://github.com/coverlet-coverage/coverlet/pull/1123/commits/356bf2702ad7c147893dcbbb7e52ac0480b616df)

File: src/coverlet.console/Program.cs [hunk 4]
```diff
@@ -174,20 +176,55 @@ static int Main(string[] args)
...
+                    }
+                }
+                
+                Dictionary<ThresholdTypeFlags, double> thresholdTypeFlagValues = new Dictionary<ThresholdTypeFlags, double>();
+                if (threshold.HasValue() && threshold.Value().Contains(','))
+                {
+                    var thresholdValues = threshold.Value().Split(',', StringSplitOptions.RemoveEmptyEntries).Select(t => t.Trim());
+                    if (thresholdValues.Count() != thresholdTypeFlagQueue.Count())
+                    {
+                        throw new Exception($"Threshold type flag count ({thresholdTypeFlagQueue.Count()}) and values count ({thresholdValues.Count()}) doesnt match");
```
Comment (daveMueller) [inline src/coverlet.console/Program.cs:206]: Could you end the exception with ```doesn't match.``` Also in ```CoverageResultTask```. Sorry for beeing petty :pray:.

File: src/coverlet.core/CoverageResult.cs [hunk 5]
```diff
@@ -111,7 +111,7 @@ internal void Merge(Modules modules)
             }
         }
 
-        public ThresholdTypeFlags GetThresholdTypesBelowThreshold(CoverageSummary summary, double threshold, ThresholdTypeFlags thresholdTypes, ThresholdStatistic thresholdStat)
+        public ThresholdTypeFlags GetThresholdTypesBelowThreshold(CoverageSummary summary, Dictionary<ThresholdTypeFlags, double> thresholdTypeFlagValues, ThresholdTypeFlags thresholdTypes, ThresholdStatistic thresholdStat)
```
Comment (daveMueller) [inline src/coverlet.core/CoverageResult.cs:114]: I guess we now could remove the ```thresholdTypes``` parameter from the method. The information it provided is now in the ```thresholdTypeFlagValues``` dictionary?

File: src/coverlet.console/Program.cs [hunk 6]
```diff
@@ -230,26 +267,25 @@ static int Main(string[] args)
                 {
                     exitCode += (int)CommandExitCodes.TestFailed;
                 }
-                thresholdTypeFlags = result.GetThresholdTypesBelowThreshold(summary, dThreshold, thresholdTypeFlags, dThresholdStat);
```
[the remark this answers is not in the document]
Comment (pbmiguel) [reply src/coverlet.console/Program.cs:233]: done. please check

File: src/coverlet.console/Program.cs (hunk 4 above)
Comment (pbmiguel) [reply src/coverlet.console/Program.cs]: no problem, changed from `doesnt match` to `doesn't match` in both files. please check

File: src/coverlet.core/CoverageResult.cs (hunk 5 above)
Comment (pbmiguel) [reply src/coverlet.core/CoverageResult.cs]: makes sense. removed `thresholdTypes` please check

File: src/coverlet.console/Program.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (daveMueller) [reply src/coverlet.console/Program.cs:233]: Thanks

File: src/coverlet.console/Program.cs (hunk 4 above)
Comment (daveMueller) [reply src/coverlet.console/Program.cs]: Thanks

File: src/coverlet.core/CoverageResult.cs (hunk 5 above)
Comment (daveMueller) [reply src/coverlet.core/CoverageResult.cs]: Thanks

File: src/coverlet.console/Program.cs [hunk 7]
```diff
@@ -174,26 +175,60 @@ static int Main(string[] args)
                 }
 
                 var thresholdTypeFlags = ThresholdTypeFlags.None;
```
Comment (daveMueller) [inline src/coverlet.console/Program.cs:177]: Sorry but I have one last request. Now that ```thresholdTypeFlags``` isn't passed to ```GetThresholdTypesBelowThreshold``` anymore. The variable here and in the following ```foreach``` can be removed.
The definition for ```thresholdTypeFlags``` can then be made in line 268 like 
```csharp
var thresholdTypeFlags = result.GetThresholdTypesBelowThreshold(summary, thresholdTypeFlagValues, dThresholdStat);
```

The same should be done in ```CoverageResultTask```.

Comment (pbmiguel) [reply src/coverlet.console/Program.cs]: removed in both places. please check @daveMueller

**Review by daveMueller — APPROVED** [review]
LGTM 👍

File: src/coverlet.console/Program.cs (hunk 7 above)
Comment (daveMueller) [reply src/coverlet.console/Program.cs]: Looks good. Thanks.

File: src/coverlet.core/CoverageResult.cs (hunk 1 above)
Comment (pbmiguel) [reply src/coverlet.core/CoverageResult.cs]: @MarcoRossignoli, can I resolve this conversation? or is there something else that must be fixed?

**Comment by pbmiguel** [comment]
@MarcoRossignoli, when will you expected this to be released? because the last release was in February

**Comment by MarcoRossignoli** [comment]
We don't have ETA yet, need to schedule, I'll take a look at issue list to understand if we want more fox before release, anyway you can dogfood our nighty build if you want, so you'll help us with testing also ☺️

**Pull request merged**